### PR TITLE
WIP: Fix Bcrypt implementation and add feature to change Key Stretching Function

### PR DESCRIPTION
**Pull request by patrickfav**
refs #16

### CODE REVIEW

**Comment by patrickfav** [comment]
@davidmigloz if you have time I would welcome a review about broken brycpt fix

**Comment by davidmigloz** [comment]
Hi Patrick! Thanks for the new implementation I'll review it today.

File: armadillo/src/main/java/at/favre/lib/armadillo/FixedBcryptKeyStretcher.java [hunk 1]
```diff
@@ -58,24 +68,23 @@
...
      * @return the Bcrypt hash of the password
      */
-    private static byte[] bcrypt(char[] password, byte[] salt, int logRounds) {
+    private static byte[] bcrypt(byte[] salt, char[] password, int logRounds) {
         StrictMode.noteSlowCall("bcrypt is a very expensive call and should not be done on the main thread");
-        return Bytes.from(BCrypt.hashpw(String.valueOf(password) + Bytes.wrap(salt).encodeHex(), generateSalt(salt, logRounds))).array();
+        byte[] passwordBytes = null;
+        try {
+            passwordBytes = charArrayToByteArray(password, StandardCharsets.UTF_8);
+            return BCrypt.withDefaults().hash(logRounds,
```
Comment (patrickfav) [inline armadillo/src/main/java/at/favre/lib/armadillo/FixedBcryptKeyStretcher.java]: I was thinking, maybe we just use the raw bcrypt hash her?
Background: Everybody implements bcrypt with the Modular Hash Format used by OpenBSD which is:

    ${version}${log-count}${16-byte-salt-radix64}{23-byte-hash-radix64}

e.g.

    $2y$06$doGnefu9cbLkJTn8sef7U.dynHJFe5hS6xp7vLWb2Zu7e8cOuMVmS

This makes sense for password hash for storage, but we provide the salt ourselves, so using the everything BUT the hash seems sloppy? Maybe there are some drawbacks I dont think about right now?

File: armadillo/src/main/java/at/favre/lib/armadillo/DefaultEncryptionProtocol.java [hunk 2]
```diff
@@ -159,6 +160,17 @@ public String deriveContentKey(String originalContentKey) {
         }
     }
 
+    @Override
+    public void setKeyStretchingFunction(KeyStretchingFunction function) {
+        Objects.requireNonNull(function);
+        keyStretchingFunction = function;
```
Comment (patrickfav) [inline armadillo/src/main/java/at/favre/lib/armadillo/DefaultEncryptionProtocol.java]: Are there any issues by being able to change the `keyStretchingFunction` after creation of the protocol?

**Comment by davidmigloz** [comment]
Sorry I couldn't review it yesterday in the end. I'll do it this evening!

File: armadillo/src/main/java/at/favre/lib/armadillo/DefaultEncryptionProtocol.java (hunk 2 above)
Comment (davidmigloz) [reply armadillo/src/main/java/at/favre/lib/armadillo/DefaultEncryptionProtocol.java]: I don't see any issue..

File: armadillo/src/main/java/at/favre/lib/armadillo/FixedBcryptKeyStretcher.java (hunk 1 above)
Comment (davidmigloz) [reply armadillo/src/main/java/at/favre/lib/armadillo/FixedBcryptKeyStretcher.java]: I agree that using the full bcrypt message has no value in this situation, as you are not storing it anywhere. I don't see any drawbacks.

**Comment by davidmigloz** [comment]
It looks very good, I haven't found anything. Good job 😄

**Comment by davidmigloz** [comment]
I was wondering whether we can have a better name for `FixedBcryptKeyStretcher` without the fixed word. What do you think?

**Comment by patrickfav** [comment]
Do you have a suggesting. I just wanted to be different from the old `BcryptKeyStretcher`

**Comment by davidmigloz** [comment]
`BCryptKeyStretcher` (with C in capital letter)? Enough to make an import from an old code fail, but still a nice name. My concern with "fixed" is that one thing is fixed until another bug is found hehe (it shouldn't be the case, but still..)

**Comment by patrickfav** [comment]
How about just ArmadilloBcryptKeyStretcher?

**Comment by davidmigloz** [comment]
Also fine 👍

**Pull request merged**